### PR TITLE
Revamped concepts doc for ReplicaSet

### PR DESCRIPTION
**Pull request by liyinan926**
This PR updated the concept doc for ReplicaSet. A few new sections were added, similarly to what the concepts docs for other workloads controllers have. This is a part of broader efforts to enhance the workloads controller docs. The corresponding example ReplicaSet YAML file was also updated to reflect the removal of selector defaulting for apps/v1beta2 in kubernetes/kubernetes#50164.

Replaces #5396. 
@kow3ns 

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/5463)

### CODE REVIEW

**Comment by kow3ns** [comment]
btw for future reference, you could have just rebased #5396 against release-1.8 and pushed to origin branch to get the update to work.

**Comment by liyinan926** [comment]
@steveperry-53. PTAL. Thanks!

**Comment by steveperry-53** [comment]
I'll do a docs review today.

**Comment by liyinan926** [comment]
@kow3ns per the discussion offline, updated with e37b45a. PTAL. Thanks!

File: docs/concepts/workloads/controllers/replicaset.md [hunk 1]
```diff
@@ -93,7 +93,99 @@ frontend-dnjpy   1/1       Running   0          1m
...
+[pod template](/docs/concepts/workloads/pods/pod-overview/#pod-templates). It has exactly the same schema as a 
+[pod](/docs/concepts/workloads/pods/pod/), except that it is nested and does not have an `apiVersion` or `kind`.
+
+In addition to required fields of a pod, a pod template in a ReplicaSet must specify appropriate
+labels and an appropriate restart policy.
+
+For labels, make sure to not overlap with other controllers. For more information, see [pod selector](#pod-selector). 
+For restart policy, only a  [`.spec.template.spec.restartPolicy`](/docs/concepts/workloads/pods/pod-lifecycle/) 
+equal to `Always` is allowed, which is the default if not specified.
+
```
Comment (steveperry-53) [inline docs/concepts/workloads/controllers/replicaset.md]: Suggestion:
The only allowed value for `.spec.template.spec.restartPolicy` is `Always`, which is the default.

File: docs/concepts/workloads/controllers/replicaset.md [hunk 2]
```diff
@@ -93,7 +93,99 @@ frontend-dnjpy   1/1       Running   0          1m
...
+For local container restarts, ReplicaSet delegate to an agent on the node,
+for example the [Kubelet](/docs/admin/kubelet/) or Docker.
+
```
Comment (steveperry-53) [inline docs/concepts/workloads/controllers/replicaset.md]: delegate to -> delegates to

File: docs/concepts/workloads/controllers/replicaset.md [hunk 3]
```diff
@@ -93,7 +93,99 @@ frontend-dnjpy   1/1       Running   0          1m
...
+The `.spec.template.metadata.labels` must match the `.spec.selector`, or it will
+be rejected by the API.
+
+In Kubernetes 1.8 or later, `.spec.selector` and `.metadata.labels` no longer default to `.spec.template.metadata.labels` if not set. So they must be set explicitly. Also note that `.spec.selector` is immutable after creation in Kubernetes 1.8 or later.
+
+Also you should not normally create any pods whose labels match this selector, either directly, with 
+another ReplicaSet, or with another controller such as Deployment. If you do so, the ReplicaSet thinks that it 
+created the other pods. Kubernetes does not stop you from doing this.
+
```
Comment (steveperry-53) [inline docs/concepts/workloads/controllers/replicaset.md]: as Deployment -> as a Deployment.

File: docs/concepts/workloads/controllers/replicaset.md [hunk 4]
```diff
@@ -93,7 +93,99 @@ frontend-dnjpy   1/1       Running   0          1m
...
+would set these the same as the `.spec.template.metadata.labels`.  However, they are allowed to be
+different, and the `.metadata.labels` do not affect the behavior of the ReplicaSet.
+
+### Replicas
+
+You can specify how many pods should run concurrently by setting `.spec.replicas` to the number
+of pods you would like to have running concurrently. The number running at any time may be higher
+or lower, such as if the replicas were just increased or decreased, or if a pod is gracefully
+shutdown, and a replacement starts early.
+
```
Comment (steveperry-53) [inline docs/concepts/workloads/controllers/replicaset.md]: shutdown -> shut down

File: docs/concepts/workloads/controllers/replicaset.md [hunk 5]
```diff
@@ -93,7 +93,99 @@ frontend-dnjpy   1/1       Running   0          1m
...
+### Labels on a ReplicaSet
+
+The ReplicaSet can itself have labels (`.metadata.labels`).  Typically, you
+would set these the same as the `.spec.template.metadata.labels`.  However, they are allowed to be
+different, and the `.metadata.labels` do not affect the behavior of the ReplicaSet.
+
+### Replicas
+
+You can specify how many pods should run concurrently by setting `.spec.replicas` to the number
+of pods you would like to have running concurrently. The number running at any time may be higher
```
Comment (steveperry-53) [inline docs/concepts/workloads/controllers/replicaset.md]: Suggestion:
Strike " to the number of pods you would like to have running cuncurrently".

File: docs/concepts/workloads/controllers/replicaset.md [hunk 6]
```diff
@@ -118,6 +209,31 @@ Alternatively, you can use the `kubectl autoscale` command to accomplish the sam
 kubectl autoscale rs frontend
 ```
 
+## Alternatives to ReplicaSet
+
+### Deployment (Recommended)
+
+[`Deployment`](/docs/concepts/workloads/controllers/deployment/) is a higher-level API object that updates its underlying Replica Sets and their Pods
+in a similar fashion as `kubectl rolling-update`. Deployments are recommended if you want this rolling update functionality,
```
Comment (steveperry-53) [inline docs/concepts/workloads/controllers/replicaset.md]: Replica Sets -> ReplicaSets

File: docs/concepts/workloads/controllers/replicaset.md [hunk 7]
```diff
@@ -118,6 +209,31 @@ Alternatively, you can use the `kubectl autoscale` command to accomplish the sam
...
+because unlike `kubectl rolling-update`, they are declarative, server-side, and have additional features. For more information on running a stateless
+application using a Deployment, please read [Run a Stateless Application Using a Deployment](/docs/tasks/run-application/run-stateless-application-deployment/).
+
+### Bare Pods
+
+Unlike in the case where a user directly created pods, a ReplicaSet replaces pods that are deleted or terminated for any reason, such as in the case of node failure or disruptive node maintenance, such as a kernel upgrade. For this reason, we recommend that you use a ReplicaSet even if your application requires only a single pod. Think of it similarly to a process supervisor, only it supervises multiple pods across multiple nodes instead of individual processes on a single node.  A ReplicaSet delegates local container restarts to some agent on the node (for example, Kubelet or Docker).
+
```
Comment (steveperry-53) [inline docs/concepts/workloads/controllers/replicaset.md]: Unlike in the case -> Unlike the case

**Comment by steveperry-53** [comment]
@liyinan926 Just a few comments in the text.

File: docs/concepts/workloads/controllers/replicaset.md (hunk 1 above)
Comment (liyinan926) [reply docs/concepts/workloads/controllers/replicaset.md]: Done.

File: docs/concepts/workloads/controllers/replicaset.md (hunk 2 above)
Comment (liyinan926) [reply docs/concepts/workloads/controllers/replicaset.md]: Done.

File: docs/concepts/workloads/controllers/replicaset.md (hunk 3 above)
Comment (liyinan926) [reply docs/concepts/workloads/controllers/replicaset.md]: Done.

File: docs/concepts/workloads/controllers/replicaset.md (hunk 4 above)
Comment (liyinan926) [reply docs/concepts/workloads/controllers/replicaset.md]: Done.

File: docs/concepts/workloads/controllers/replicaset.md (hunk 6 above)
Comment (liyinan926) [reply docs/concepts/workloads/controllers/replicaset.md]: Done.

File: docs/concepts/workloads/controllers/replicaset.md (hunk 7 above)
Comment (liyinan926) [reply docs/concepts/workloads/controllers/replicaset.md]: Done.

**Pull request merged**